File: src/main/java/no/fintlabs/mapping/SakMappingService.java
package no.fintlabs.mapping;

import no.fint.model.resource.Link;
import no.fint.model.resource.arkiv.noark.SakResource;
import no.fintlabs.model.instance.SakDto;
import org.springframework.stereotype.Service;

@Service
public class SakMappingService {

    private final SkjermingMappingService skjermingMappingService;
    private final KlasseMappingService klasseMappingService;
    private final PartMappingService partMappingService;

    public SakMappingService(
            SkjermingMappingService skjermingMappingService,
            KlasseMappingService klasseMappingService,
            PartMappingService partMappingService) {
        this.skjermingMappingService = skjermingMappingService;
        this.klasseMappingService = klasseMappingService;
        this.partMappingService = partMappingService;
    }

    public SakResource toSakResource(SakDto sakDto) {
        if (sakDto == null) {
            return null;
        }
        SakResource sakResource = new SakResource();
        sakDto.getTittel().ifPresent(sakResource::setTittel);
        sakDto.getOffentligTittel().ifPresent(sakResource::setOffentligTittel);
        sakDto.getSaksmappetype().map(Link::with).ifPresent(sakResource::addSaksmappetype);
        sakDto.getSaksstatus().map(Link::with).ifPresent(sakResource::addSaksstatus);
        sakDto.getJournalenhet().map(Link::with).ifPresent(sakResource::addJournalenhet);
        sakDto.getAdministrativenhet().map(Link::with).ifPresent(sakResource::addAdministrativEnhet);
        sakDto.getSaksansvarlig().map(Link::with).ifPresent(sakResource::addSaksansvarlig);
        sakDto.getArkivdel().map(Link::with).ifPresent(sakResource::addArkivdel);
        sakDto.getPart()
                .map(part -> part
                        .stream()
                        .map(partMappingService::toPartResource)
                        .toList()
                )
                .ifPresent(sakResource::setPart);
        sakDto.getSkjerming()
                .map(skjermingMappingService::toSkjermingResource)
                .ifPresent(sakResource::setSkjerming);
        sakDto.getKlasse()
                .map(klasseMappingService::toKlasse)
                .ifPresent(sakResource::setKlasse);

        return sakResource;
    }

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/dispatch/mapping/SakMappingService.kt
package no.novari.flyt.archive.gateway.dispatch.mapping

import no.novari.fint.model.resource.Link
import no.novari.fint.model.resource.arkiv.noark.SakResource
import no.novari.flyt.archive.gateway.dispatch.model.instance.SakDto
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class SakMappingService(
    private val skjermingMappingService: SkjermingMappingService,
    private val klasseMappingService: KlasseMappingService,
    private val partMappingService: PartMappingService,
) {
    fun toSakResource(sakDto: SakDto?): SakResource {
        requireNotNull(sakDto) { "sakDto cannot be null" }

        log.info("Mapping SakDto to SakResource")

        return SakResource()
            .apply {
                sakDto.tittel?.let(::setTittel)
                sakDto.offentligTittel?.let(::setOffentligTittel)
                sakDto.saksmappetype?.let(Link::with)?.let(::addSaksmappetype)
                sakDto.saksstatus?.let(Link::with)?.let(::addSaksstatus)
                sakDto.tilgangsgruppe?.let(Link::with)?.let(::addTilgangsgruppe)
                sakDto.journalenhet?.let(Link::with)?.let(::addJournalenhet)
                sakDto.administrativEnhet?.let(Link::with)?.let(::addAdministrativEnhet)
                sakDto.saksansvarlig?.let(Link::with)?.let(::addSaksansvarlig)
                sakDto.arkivdel?.let(Link::with)?.let(::addArkivdel)
                sakDto.part?.mapNotNull(partMappingService::toPartResource)?.let(::setPart)
                sakDto.skjerming?.let(skjermingMappingService::toSkjermingResource)?.let(::setSkjerming)
                sakDto.klasse?.let(klasseMappingService::toKlasse)?.let(::setKlasse)
            }.also {
                log.info("Successfully mapped SakDto to SakResource")
            }
    }

    companion object {
        private val log = LoggerFactory.getLogger(SakMappingService::class.java)
    }
}
